Interactive TX negotiation tracks shared outputs

optout21 opened the pull request:
Interactive TX negotiation now tracks the shared funding output. This allows for better checks, better contribution verification.
Shared inputs are not included, those are splicing-specific, shall be done later.
Besed on https://github.com/lightningdevkit/rust-lightning/commit/36e74526718773eb6c312aa994cfd4881a4c827d by @dunxen .

One notable change is that local/remote contribution are not computed by filtering by serial_id and adding values, but summing local/remote values across all inputs/outputs (i.e. in `build_transaction()` https://github.com/lightningdevkit/rust-lightning/pull/2989/files#diff-c92e994fd9949ebcb6a8d9ec76553e8497ee223b707b51905e35017b80124c77R467).

~~Housekeeping: this should come after #2981~~

dunxen commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -85,15 +87,105 @@ struct NegotiationContext {
 	received_tx_add_output_count: u16,
-	inputs: HashMap<SerialId, TxInputWithPrevOutput>,
+	/// The inputs to be contributed by the holder.
+	inputs: HashMap<SerialId, InteractiveTxInput>,
```
I've actually thought it would be better to split this into local and remote maps as quite often we're iterating/checking either local / remote. The only time they're combined is during validating and building the final transaction.

I have this change locally for #2981 which I'm pushing up today.

dunxen replied:
Same with the outputs.

optout21 commented:
Changed `NegotiationContext` to store local and remote inputs/output in separate vectors (in separate commit for now)

dunxen commented:
> Changed `NegotiationContext` to store local and remote inputs/output in separate vectors (in separate commit for now)

Sorry about this, but I've reverted this in #2423 as there just ended up being to much cloning going around. I believe we will still be fine with the splicing additions here to have single maps for inputs/outputs.

optout21 commented:
Rebased after merge of #2981 .

dunxen commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -822,6 +870,126 @@ impl StateMachine {
+pub enum InteractiveTxInput {
+	Local(LocalOrRemoteInput),
+	Remote(LocalOrRemoteInput),
+	// Note: SharedInput should be added later, with splicing
```
Noted.

nit: Could we make this a `TODO(splicing):` comment?

dunxen commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -132,18 +120,12 @@ impl ConstructedTransaction {
-				!is_serial_id_valid_for_counterparty(context.holder_is_initiator, serial_id)
-			})
-			.fold(0u64, |value, (_, input)| value.saturating_add(input.prev_output.value));
+			.fold(0u64, |value, (_, input)| value.saturating_add(input.local_value()));
```
This is nice to have the input/output variants back again 👍

dunxen commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -201,10 +175,32 @@ struct NegotiationContext {
 	holder_is_initiator: bool,
 	received_tx_add_input_count: u16,
 	received_tx_add_output_count: u16,
+	/// The inputs
```
nit: Seems unnecessary?

dunxen commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -201,10 +175,32 @@ struct NegotiationContext {
 	received_tx_add_output_count: u16,
+	/// The inputs
 	inputs: HashMap<SerialId, InteractiveTxInput>,
+	/// The output intended to be the new funding output.
+	/// When an output added to the same pubkey, it will be treated as the shared output.
+	/// The script pubkey is used to discriminate which output is the funding output.
+	/// The value provided here is just a hint, when the actual output is added,
+	/// its value will take over.
+	intended_new_funding_output: TxOut,
+	/// The holder's intended contribution from the shared funding output,
+	/// in the range 0 -- intended_new_funding_output.value.
+	/// The rest is the counterparty's contribution
+	/// (intended_new_funding_output.value - intended_local_contribution_satoshis)
+	/// When the funding output is added (recognized by its output pubkey), it will be marked
+	/// as shared, split between the peers according to this value.
+	/// (In the rare case of actual output value differing from intended, the intended local
+	/// contribution may be clipped.)
+	intended_local_contribution_satoshis: u64,
+	/// The actual new funding output, set only after the output has been actually added.
+	/// Note: this output is also included in `outputs`.
```
some small nits
```suggestion
	/// The output intended to be the new funding output.
	/// When an output is added with the same `script_pubkey`, it will be treated as the shared output.
	/// The `script_pubkey` field is used to determine which output is the funding output.
	/// The `value` provided here is just a hint, when the actual output is added,
	/// its `value` will take over.
	intended_new_funding_output: TxOut,
	/// The holder's intended contribution to the shared funding output,
	/// in the range [0, intended_new_funding_output.value].
	/// The rest is the counterparty's contribution
	/// (intended_new_funding_output.value - intended_local_contribution_satoshis)
	/// When the funding output is added (recognized by its output `script_pubkey`), it will be marked
	/// as shared, and split between the peers according to this value.
	/// (In the rare case of actual output value differing from intended, the intended local
	/// contribution may be clipped.)
	intended_local_contribution_satoshis: u64,
	/// The actual new funding output, set only after the output has actually been added.
	/// NOTE: This output is also included in `outputs`.
```

dunxen reviewed (commented):
Apologies for the delay. I've done some initial review and so far LGTM. Did you want to tackle splicing calculations in a separate PR? I was thinking it would be okay to do that here and track the shared inputs too. Or do imagine that to be quite involved and more than a few extra commits?

Either way this looks good to take out of draft if you're happy and I'll give it another round once you're happy 🙂

optout21 commented:
> Apologies for the delay. I've done some initial review and so far LGTM. Did you want to tackle splicing calculations in a separate PR? I was thinking it would be okay to do that here and track the shared inputs too. Or do imagine that to be quite involved and more than a few extra commits?

With current state of splicing work, I still don't see exactly the way (and the need) to track the shared output, so I would defer that to the time splicing is added.

optout21 commented:
Folded in proposed comment changes; marked ready for review.

dunxen commented:
> Folded in proposed comment changes; marked ready for review.

Thanks. I'll give this another look over today.

dunxen commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -430,6 +463,14 @@ impl NegotiationContext {
 		}
 
+		let txout = TxOut { value: msg.sats, script_pubkey: msg.script.clone() };
+		let output = if msg.script == self.intended_new_funding_output.script_pubkey {
```
Initiator always sends the shared output (and input in the case of splicing) but there is no restriction on when it is sent.

In the case of splicing, there is an explicit tlv for the funding txid which identifies it as the shared input. A funding output just looks like a regular tx_add_output so needs to be identified.

dunxen replied:
Yeah it's still strange...

optout21 commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -202,9 +176,30 @@ struct NegotiationContext {
+	/// The `script_pubkey` field is used to determine which output is the funding output.
+	/// The `value` provided here is just a hint, when the actual output is added,
+	/// its `value` will take over.
+	intended_new_funding_output: TxOut,
```
Indeed, this can be removed, it is a bit of a simplification.

dunxen reviewed (dismissed):
LGTM. I'm currently writing some more tests at the channelmanager-level to cover some more interesting funding cases.

Looking forward to rebasing on this.

optout21 commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -233,35 +224,70 @@ fn is_serial_id_valid_for_counterparty(holder_is_initiator: bool, serial_id: &Se
 		is_serial_id_valid_for_counterparty(self.holder_is_initiator, serial_id)
 	}
 
+	fn total_input_and_output_count(&self) -> usize {
```
Indeed, removed   (dead code warning is turned off for the whole file)

optout21 commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -464,32 +495,41 @@ impl NegotiationContext {
+			input: txin,
+			prev_output: msg
+				.prevtx
+				.as_transaction()
```
Done

optout21 commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -233,35 +224,70 @@ fn is_serial_id_valid_for_counterparty(holder_is_initiator: bool, serial_id: &Se
+
+		let value = txout.value;
+		// clamp local contribution if needed
+		let local_value = cmp::min(value, self.intended_local_contribution_satoshis);
```
First, this pathway is executed regardless who adds the shared funding output.
If the same node adds it, and the intended local part is higher, that's inconsistent input from the node.
If the other node adds it, that's an interop error.
In either case it makes sense to abort, indeed.
I changed that (plus added unit tests).

optout21 commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -876,13 +1036,19 @@ impl InteractiveTxConstructor {
 		inputs_to_contribute: Vec<(TxIn, TransactionU16LenLimited)>,
-		outputs_to_contribute: Vec<TxOut>,
+		outputs_to_contribute: Vec<TxOut>, intended_new_funding_output_script: ScriptBuf,
+		intended_local_contribution_satoshis: u64,
```
Not really, because of fees. The share of the output cannot be higher than the share of the inputs, but it can be lower, the rest being used for fees (fees have to meet a minimum).

optout21 commented:
Made some smaller changes according to review comments.

optout21 commented:
The last review questions made me thinking, if the API can be made more straightforward.

The user of the `InteractiveTxConstructor` has to supply the provided outputs (`TxOut`'s), plus it has to provide the script pubkey of the expected shared output (one of the outputs), and its contribution from that. This setup has a few easy places for inconsistency.

Maybe it would be better to wrap the provided outputs in an enum, for different types:
- shared output, but its value belonging entirely to initiator (typical funding output)
- shared output, with partial value belonging initiator (this value has to be provided, this is a less typical case)
- output belonging to initiator (e.g. change)

Additionally, for the infrequent case when we don't specify any outputs, but expect a shared output and some balance belonging to us, an optional (pubkey+localvalue) should be provided.

Maybe this way the API would be easier to understand.
I put the PR back to Draft, for a 2nd thought.

optout21 commented:
Reworked:
- In the `InteractiveTxConstructor::new()`, the type of the contributed outputs has to be specified -- controlled by local node, shared output but all belongs to the local node, or shared output with joint ownership (split).
- Additional expected shared output script has to be specified only in the rare case when the local node does not add a shared output, but expects the remote node to add a joint-owned shared output.

optout21 commented:
> Rather than pushing type info upstream, maybe we should have simpler (public) constructors for the `InteractiveTxConstructor` that are context-specific? ie `InteractiveTxConstructor::for_splicing` and `InteractiveTxConstructor::for_dual_funding` are kinda separate things, and could use separate constructors.

Good idea, though this version aims at dual funding only (only funding outputs), splicing is not considered in this PR

dunxen commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -822,11 +865,170 @@ impl StateMachine {
+
+/// Represents an output, with information about
+/// its control -- exclusive by the adder or shared --, and
+/// its ownership -- fully owned by the adder or jointly
```
I think we need to expand on the meaning of "owned" here. I assume it relates to the value. So "value fully owned" might be clearer at least in the docs. The variant naming can remain.

optout21 replied:
Yes, ownership here refers to whom the satoshis belong. Done.

dunxen commented:
I've been writing some new tests for #2302 and have encountered some issues due to not tracking the shared output within the constructor and when validating the constructed tx, so this PR is a little bit of a blocker for 2302. I'll integrate these changes for now though so I can carry on with the new tests. :)

optout21 commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -184,13 +164,11 @@ impl ConstructedTransaction {
-			inputs.into_iter().map(|InteractiveTxInput { input, .. }| input).collect();
-		let output: Vec<TxOut> =
-			outputs.into_iter().map(|InteractiveTxOutput { tx_out, .. }| tx_out).collect();
+		let input: Vec<TxIn> = inputs.into_iter().map(|input| input.txin().clone()).collect();
+		let output: Vec<TxOut> = outputs.into_iter().map(|output| output.tx_out().clone()).collect();
```
I couldn't really. Inside the the map function we get references to `TxOut` and `TxIn` obejcts, which don't to `Copy`, and we want to provide them to the `Transaction` constructor.

optout21 commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -202,9 +180,25 @@ struct NegotiationContext {
+	expected_shared_funding_output: Option<(ScriptBuf, u64)>,
+	/// The actual new funding output, set only after the output has actually been added.
+	/// NOTE: this output is also included in `outputs`.
+	actual_new_funding_output: Option<SharedOwnedOutput>,
```
That makes sense. It affects many unit tests (that don't set a shared output). Will do.

optout21 replied:
Done.

optout21 commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -202,9 +180,25 @@ struct NegotiationContext {
+	expected_shared_funding_output: Option<(ScriptBuf, u64)>,
+	/// The actual new funding output, set only after the output has actually been added.
+	/// NOTE: this output is also included in `outputs`.
+	actual_new_funding_output: Option<SharedOwnedOutput>,
```
The point here is that the user of `NegotiationContext` will be able to find the funding output from this field, without the need to iterate through the outputs. The whole `NegotiationContext` struct is unused in main, therefore this fields also looks unused.

optout21 commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -233,35 +227,63 @@ fn is_serial_id_valid_for_counterparty(holder_is_initiator: bool, serial_id: &Se
+			return Err(AbortReason::DuplicateFundingOutput);
+		}
+
+		assert!(self.expected_shared_funding_output.is_some());
```
This method is called only when `expected_shared_funding_output` is set, and I added the `assert` to indicate that.
But it is more robust with no assert; I added an if clause instead -- with an unreachable error branch.

optout21 commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -870,19 +1072,62 @@ pub(crate) enum HandleTxCompleteValue {
-		outputs_to_contribute: Vec<TxOut>,
-	) -> (Self, Option<InteractiveTxMessageSend>)
+		outputs_to_contribute: Vec<OutputOwned>,
+		expected_remote_shared_funding_output: Option<(ScriptBuf, u64)>,
```
It is only needed when the local node will not add the funding tx, but it expects the remote node to add it.
I think it's used by the acceptor node.

optout21 replied:
Clatification:
- `expected_remote_shared_funding_output` (an input to `InteractiveTxConstructor::new()`) is needed to be set, when the node does not provide a funding output, but expects the counterparty to do so.
- `expected_shared_funding_output` (an input to `NegotiationContext::new()`) is mandatory now. It either comes from the shared output provided or the expected remote funding output (but one has to be there).

optout21 commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -870,19 +1072,62 @@ pub(crate) enum HandleTxCompleteValue {
+				// one local-added and one remote-expected shared output
+				return Err(AbortReason::DuplicateFundingOutput);
+			}
+			shared_funding_output = Some(expected_remote_shared_funding_output.unwrap());
```
No, here the logic is to find the locally to-be-added shared output, or the expected-added remote shared output, and at most one can be. The local output we have (they will be added), but the remote ones we don't yet, so no comparison can be made here.

optout21 commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -464,32 +496,44 @@ impl NegotiationContext {
+		let input = InteractiveTxInput::Local(LocalOrRemoteInput {
+			serial_id: msg.serial_id,
+			input: txin,
+			prev_output: tx
```
Done

optout21 commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -870,19 +1072,62 @@ pub(crate) enum HandleTxCompleteValue {
+					// more than one local-added shared output
+					return Err(AbortReason::DuplicateFundingOutput);
+				}
+				shared_funding_output = new_output;
```
I don't fully understand what you suggest, but I simplified the checks a bit.

optout21 commented:
Addressed review comments (sorry, they fall of my priority table...).
One comment is still open: whether there should be a check in `validate_tx` for the existence of a shared output.

optout21 commented:
Did remaining changes:

- Add check for missing funding output in validate_tx()
- Make `expected_shared_funding_output` mandatory in `NegotiationContext`, as suggested by @TheBlueMatt . In `InteractiveTxConstructor` either a shared output has to be provided, or a pubkey for the shared output to be added by the remote.

dunxen commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -879,57 +1066,101 @@ pub(crate) enum HandleTxCompleteValue {
+	/// `expected_remote_shared_funding_output`: In the case when the local node doesn't
+	/// add a shared output, but it expects a shared output to be added by the remote node,
+	/// it has to specify the script pubkey, used to determine the shared output,
+	/// and its (local) contribution from the shared output:
+	/// 0 when the whole value belongs to the remote node, positive if owned also by local.
+	/// The local value cannot be larger that the actual shared output.
+	///
```
```suggestion
	/// and its (local) contribution from the shared output:
	///   0 when the whole value belongs to the remote node, or
	///   positive if owned also by local.
	/// Note: The local value cannot be larger than the actual shared output.
	///
```

dunxen commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -879,57 +1066,101 @@ pub(crate) enum HandleTxCompleteValue {
-			StateMachine::new(feerate_sat_per_kw, is_initiator, funding_tx_locktime);
-		let mut inputs_to_contribute: Vec<(SerialId, TxIn, TransactionU16LenLimited)> =
-			inputs_to_contribute
+		// Sanity check: There can be at most one shared output, local-added or remote added
```
```suggestion
		// Sanity check: There can be at most one shared output, local-added or remote-added
```

dunxen reviewed (commented):
Just some comment nits, but LGTM. I believe this needs to be formatted with `rustfmt` as it's not excluded.

optout21 commented:
Did formatting, plus 2 minor comment wording change (review).

optout21 commented:
If this is merged before #3137, these interactivetx.rs changes will disappear from #3137 diff. That would make sense, as this PR has been mostly reviewed already.

dunxen commented:
> If this is merged before #3137, these interactivetx.rs changes will disappear from #3137 diff. That would make sense, as this PR has been mostly reviewed already.

Yip, that's the plan. I just cherry-picked this.

dunxen commented:
Since this only affects `interactivetxs.rs`, I'm going to go ahead and merge this and any other nits/issues can be addressed in a follow-up. Then that should unblock 3137.

optout21 commented on lightning/src/ln/interactivetxs.rs:
```diff
@@ -831,11 +877,182 @@ impl StateMachine {
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum InteractiveTxInput {
```
Looks like visibility could be reduced indeed. I will check in the wider scope where this class is actually used.
These are used in `ConstructedTransaction`, which is `pub(crate)`.
